src/ryu.jl

```diff
@@ -41,39 +44,70 @@ function format_fixed_scientific(x::AbstractFloat, precision::Integer,
     end
 
     if engineering
-        b, e10 = Ryu.reduce_shortest(convert(Float32, x))
-        d, r = divrem(e10, 3)
-        if d < 0 &&
-            d += sign(r)
-
+        base_digits, power = get_engineering_string(x, precision)
+    else
+        e_format_number = Ryu.writeexp(x, precision)
+        base_digits, power = split(e_format_number, 'e')
     end
-    ryustr = Ryu.writeexp(x, precision)
-    @show x ryustr
 
-    # Rewrite the exponent
+
     buf = IOBuffer()
-    ret = iterate(ryustr)
-    while ret !== nothing
-        c, state = ret
-        c === 'e' && break
-        print(buf, c)
-        ret = iterate(ryustr, state)
+
+    print(buf, base_digits)
+    print(buf, "×10")
+
+    if power[1] == '-'
+        print(buf, '⁻')
     end
-    if ret !== nothing
-        print(buf, "×10")
-        _, state = ret
-        ret = iterate(ryustr, state)
-        while ret !== nothing
-            c, state = ret
-            if '0' <= c <= '9'
-                print(buf, superscript_numerals[c - '0' + 1])
-            elseif c == '-'
-                print(buf, '⁻')
-            end
-            ret = iterate(ryustr, state)
+    leading_index = findfirst(c -> '1' <= c <= '9', power)
+
+    if isnothing(leading_index)
```

Comment:
```suggestion
    if leading_index === nothing
```
is friendlier for inference (with your version, it won't know that in `for digit in power[leading_index:end]` that `leading_index` must be an `Int`). xref https://github.com/JuliaLang/julia/issues/37342